### DISCUSSION TITLE
Metaclass inheritance with 'type(SomeClass)'

### DISCUSSION THREAD

**Discussion by vladislav-krastev** [microsoft/pyright#9452] [Q&A] [answered]
Maybe I'm missing something, so that's why I didn't rush reporting it as a bug.

Using VSCode:
**Pylance _v2024.11.1_**
**pyright _v1.1.386_** (from how I understand it [Pylance CHANGELOG](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202411100-6-november-2024-prerelease))

Running the following (simplest) example:
```python
class MyBaseMeta(type):
    def __new__(cls, name: str, bases: tuple, namespace: dict):
        print("in MyBaseMeta.__new__()")
        return super().__new__(cls, name, bases, namespace)

class MyClass_1(metaclass=MyBaseMeta):
    pass

class MyChildMeta(MyBaseMeta):
    pass

class MyClass_2(metaclass=MyChildMeta):
    pass
```
correctly prints `in MyBaseMeta.__new__()` two times AND no issues reported from the type-checker.

However, changing **_only_** the declaration of  `MyChildMeta` to 
```python
class MyChildMeta(type(MyClass_1)):
    pass
```
still correctly prints `in MyBaseMeta.__new__()` two times BUT the type-checker reports the error `Class cannot derive from itself`.

The type-error-producing code seems correct Python to me, so is anyone having some explanation why I'm getting this message?

For context, the **BaseMeta** class is an internal class in a 3rd party library, and the **MyClass_1** is a public class of that library, so I have pretty-looking access to **MyClass_1**, but not to the **BaseMeta** class, that's why I'm trying to inherit from **BaseMeta** with `type(MyClass_1)` and not directly.

Thanks in advance!

**Comment by erictraut** [comment, marked as answer]
This issue was recently reported here: https://github.com/microsoft/pyright/issues/9430. This is addressed in pyright 1.1.389, which I just published. With this fix in place, your code above type checks without any errors.

Code sample in [pyright playground](https://pyright-play.net/?code=MYGwhgzhAECyCeAhSBTWKAuYAUH4AcUBKALgChpLoATFAM2gH1GA7FAd2e1AgBpoWYALYoS0CBgBO-AEaoIYjAFd8IFP0EiI%2BMMFE0AlsAykKVc-kkGWGbACJrcJKnRYAdMzadG2InaJm5pSSmEqSLOIqKJK%2BHqwcXDwawurQchAofAIp2rrEZAWgkDAIAMLgUIwAjNgiWEVQALwIyBmuYKYWxQVkDSXwpQAWBiDU7bgEKNhlFRDVRJ1UOlA9fU7lxYwATLWYurPNA8Oj7YuUyxBkQA)

This fix will appear in a future release of pylance.

**Reply by vladislav-krastev** [reply]
I was even talking exactly about pydantic's **BaseModel** and **ModelMetaclass**...
Apologies for missing the issue and duplicating.
Thank you!!!
